File: src/apis/log.ts
import { requestAsync } from ".";


enum Action {
  Debug = "debug",
  Error = "error",
  Info = "info",
  Trace = "trace",
  Warn = "warn",
}

class Log {
  sendRequest = async (action: Action, message: any) => {
    const payload: any = {
      target: "UI",
      message,
    };

    try {
      await requestAsync("log", action, payload);
    } catch (e) {
      console.log("send log failed: ", e);
    }
  };

  debug = (message: any) => {
    this.sendRequest(Action.Debug, message);
  };

  error = (message: any) => {
    this.sendRequest(Action.Error, message);
  };

  info = (message: any) => {
    this.sendRequest(Action.Info, message);
  };

  trace = (message: any) => {
    this.sendRequest(Action.Trace, message);
  };

  warn = (message: any) => {
    this.sendRequest(Action.Warn, message);
  };
}

export default new Log();

File: src-tauri/src/handlers/log.rs

use std::time::{Instant};


use serde::{Deserialize, Serialize};
use serde_json::Value;

use super::{Endpoint, Response, ResponseResult};

#[derive(Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub enum Action {
    Debug,
    Error,
    Info,
    Trace,
    Warn,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Payload {
    target: Option<String>,
    message: Value,
}

pub fn handle_command(endpoint: Endpoint<Action, Payload>) -> Response<()> {
    let Endpoint { action, payload } = endpoint;
    let Payload { target, message } = payload;
    let now = Instant::now();


    let level = match action {
        Action::Debug => log::Level::Debug,
        Action::Error => log::Level::Error,
        Action::Info => log::Level::Info,
        Action::Trace => log::Level::Trace,
        Action::Warn => log::Level::Warn,
    };

    match target {
        Some(tar) => log::log!(target: &tar, level, "{}", message),
        None => log::log!(level, "{}", message),
    }

    Response::new(true, now.elapsed(), ResponseResult::Result(()))
}
